### PR TITLE
Fill out section on trusted UI

### LINKED ISSUE #718: What is Trusted Immersive UI?
**[Disclaimer: This issue is one of several being filed to capture discussions that began either on #638, on #689, or at the most recent F2F]**

The concept of “Trusted UI” is what allows User Agents to display a UI to end users on which sensitive information can be displayed and interacted with such that a website cannot snoop on it and cannot spoof it. Some features which use Trusted UI are user consent prompts, URL bars, navigation controls, favorite/bookmarks, and many more.

In 2D browsers, Trusted UI is presented either exclusively around the outside of a web page’s visual container or overlapping with it partially. In the context of an immersive experience, the definition of a “Trusted Immersive UI” is a bit more complex due to the fact there is no “outside” of immersive content; all pixels the user sees are rendered by the immersive content.

Various User Agents are exploring variations of Trusted Immersive UI, and as such WebXR should probably have a stance on the minimum requirements of these UIs to ensure they can meet an agreed upon security bar. For example, one already agreed upon aspect of WebXR is that there must always be a hardware button or dedicated gesture that users can rely on to bring them out of an immersive browsing session.

This GitHub issue is to track the discussion of what this bar should be. These are some of the ideas I recall folks mentioning:

* On devices with motion controllers with buttons, a dedicated button press/hold could be used to pull up the Trusted UI. When a trusted UI popup is needed, interacting with the same dedicated hardware button before displaying the content could be required. (e.g. “This browser needs your attention. Please press the Application button.”). Alternatively, it could go straight into presenting the request but require the dedicated hardware button to make a selection (e.g. “This site is requesting your permission to blahdiblah. Press the Application button to see your choices”)
* Displaying a sigil/totem that the User Agent knows about but the website does not. This approach would probably require disabling MR video and photo capture while the sigil is displayed.
* Regardless of how the UI is presented, frame data will need to be modified in some way (significantly throttled, quantized, rounded, etc) when Trusted Immersive UI is displayed to ensure the site cannot sniff password or URL entry.

I’m sure I’ve forgotten some and there are likely others not mentioned yet. So please please chime in with any information you can share about the approach your products are taking, and from there we’ll work on distilling the the information and establishing a consistent baseline.

### LINKED ISSUE #719: Is Trusted Immersive UI a conformance requirement?
**[Disclaimer: This issue is one of several being filed to capture discussions that began either on #638, on #689, or at the most recent F2F]**

This issue is NOT to discuss the requirements for what Trusted Immersive UI should look or behave like. That conversation is happening in #718

Instead, this issue tracks the question of whether a Trusted Immersive UI is a requirement for a UA to be considered conformant with WebXR.

(Note: this is also related to issue #702 which discusses figuring out what to do with non-XR feature permissions which developers may request during an immersive session)

### PR DESCRIPTION
**Pull request by Manishearth**
/fixes #718, /fixes #719

Unsure of this approach, still working on things.

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2214,25 +2214,39 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+A [=trusted UI=] which does not exit immersive mode is known as a <dfn>trusted immersive UI</dfn>. User agents are not required to support [=trusted immersive UI=], they may instead temporarily pause/exit immersive mode to handle prompts to the user.
+
+<div class="note">
+Note: Examples of [=trusted UI=] include:
+  - The default 2D mode browser shown when not in immersive mode
+  - A prompt shown within immersive mode which can only be interacted with via a reserved hardware button to prevent spoofing
+  - Pausing the immersive session and showing some form of desktop environment in which a prompt can be shown
+
+</div>
+
```
Comment (avadacatavra) [inline index.bs]: Did we also want to enumerate the properties of a trusted UI from https://github.com/immersive-web/webxr/issues/718

**Comment by Manishearth** [comment]
Ah, yes I definitely should do that, thanks

File: index.bs (hunk 1 above)
Comment (Manishearth) [reply index.bs]: done!

File: index.bs [hunk 2]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+    - It must not be spoofable
+    - It indicates where the request/content displayed originates from
+    - It doesn't rely on a shared secret that can be observed by a mixed reality capture (e.g. a gesture that can be seen by the camera)
+    - It is consistent between immersive experiences in the same UA
+
+<div class="note">
+Note: Examples of [=trusted UI=] include:
+  - The default 2D mode browser shown when not in immersive mode
+  - A prompt shown within immersive mode which can only be interacted with via a reserved hardware button to prevent spoofing
+  - Pausing the immersive session and showing some form of desktop environment in which a prompt can be shown
```
Comment (toji) [inline index.bs]: > Pausing the immersive session and showing some form of desktop environment in which a prompt can be shown

I feel like the phrasing on this one is maybe a tad bit too restrictive? Specifically the use of the phrase "desktop environment", which seems to suggest a very specific and potentially mobile-excluding concept.

What about something along the lines of this?

>  Pausing the immersive session to show a prompt in the native system environment.

Comment (Manishearth) [reply index.bs]: done

File: index.bs [hunk 3]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
 
 Trusted Environment {#trustedenvironment-security}
 -------------------
 
-If the virtual environment does not consistently track the user's head motion with low latency and at a high frame rate the user may become disoriented or physically ill. Since it is impossible to force pages to produce consistently performant and correct content the user agent MUST provide a tracked, trusted environment and an [=XR Compositor=] which runs asynchronously from page content. The compositor is responsible for compositing the trusted and untrusted content. If content is not performant, does not submit frames, or terminates unexpectedly the user agent should be able to continue presenting a responsive, trusted UI.
+The user agent MUST support showing a <dfn>Trusted UI</dfn>, that is, an interface that the user can trust comes from the user agent, which the user may interact with without interference from the page. Some form of [=trusted UI=] MUST be used to show permissions prompts.
```
Comment (NellWaliczek) [inline index.bs]: It might be a good idea to put a couple sentences before this elaborating on why trusted UI is a harder problem in immersive experiences (i.e. all pixels can be drawn by the developer which allows for spoofing).

File: index.bs [hunk 4]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+
+Additionally, page content has the ability to make users uncomfortable in ways not related to performance. Badly applied tracking, strobing colors, and content intended to offend, frighten, or intimidate are examples of content which may cause the user to want to quickly exit the XR experience. Removing the XR device in these cases may not always be a fast or practical option. To accommodate this the user agent SHOULD provide users with an action, such as pressing a reserved hardware button or performing a gesture, that escapes out of WebXR content and displays the user agent's [=trusted UI=].
 
 {{XRSession}}s MUST have their [=visibility state=] set to {{XRVisibilityState/"hidden"}} when the user is interacting with potentially sensitive UI from the user agent (such as entering a URL) in the trusted environment.
```
Comment (NellWaliczek) [inline index.bs]: Is this a duplicate of the last paragraph in the gaze tracking section?

File: index.bs (hunk 3 above)
Comment (NellWaliczek) [inline index.bs]: Nit: there's something about this sentence that seems somewhat oddly phrased, but I can't put my finger on how it might be clearer.
>The user agent MUST support showing a <dfn>Trusted UI</dfn>, that is, an interface that the user can trust comes from the user agent, which the user may interact with without interference from the page.

Comment (NellWaliczek) [inline index.bs]: >Some form of [=trusted UI=] MUST be used to show permissions prompts.

Double checking... this section is normative, yes?

File: index.bs [hunk 5]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
 
-Additionally, page content has the ability to make users uncomfortable in ways not related to performance. Badly applied tracking, strobing colors, and content intended to offend, frighten, or intimidate are examples of content which may cause the user to want to quickly exit the XR experience. Removing the XR device in these cases may not always be a fast or practical option. To accommodate this the user agent SHOULD provide users with an action, such as pressing a reserved hardware button or performing a gesture, that escapes out of WebXR content and displays the user agent's trusted UI.
 
-When navigating between pages in XR the user agent should display trusted UI elements informing the user of the security information of the site they are navigating to which is normally presented by the 2D UI, such as the URL and encryption status.
+A [=trusted UI=] which does not exit immersive mode is known as a <dfn>trusted immersive UI</dfn>. User agents are not required to support [=trusted immersive UI=], they may instead temporarily pause/exit immersive mode to handle prompts to the user.
```
Comment (NellWaliczek) [inline index.bs]: It would probably be helpful to state up front that there are two options for user agents to choose.

File: index.bs [hunk 6]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+
+A [=trusted UI=] MUST have the following properties:
+
+
+    - It must not be spoofable
+    - It indicates where the request/content displayed originates from
+    - It doesn't rely on a shared secret that can be observed by a mixed reality capture (e.g. a gesture that can be seen by the camera)
```
Comment (NellWaliczek) [inline index.bs]: > If it relies on a shared secret with the user, this shared secret cannot be observed by a mixed reality capture (e.g. it may not be a gesture that can be seen by the camera)

File: index.bs [hunk 7]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+    - It is consistent between immersive experiences in the same UA
+
+<div class="note">
+Note: Examples of [=trusted UI=] include:
```
Comment (NellWaliczek) [inline index.bs]: Given that the text immediately above is specifically talking about trusted immersive ui, it took me a moment to realize this was back to being examples of the more general definition.

File: index.bs [hunk 8]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+  - The default 2D mode browser shown when not in immersive mode
+  - A prompt shown within immersive mode which can only be interacted with via a reserved hardware button to prevent spoofing
+  - Pausing the immersive session and showing some form of native system environment in which a prompt can be shown
+
+</div>
+
+
+
+If the virtual environment does not consistently track the user's head motion with low latency and at a high frame rate the user may become disoriented or physically ill. Since it is impossible to force pages to produce consistently performant and correct content the user agent MUST provide a tracked, trusted environment and an [=XR Compositor=] which runs asynchronously from page content. The compositor is responsible for compositing the trusted and untrusted content. If content is not performant, does not submit frames, or terminates unexpectedly the user agent should be able to continue presenting a responsive, [=trusted UI=].
```
Comment (NellWaliczek) [inline index.bs]: This seems unrelated enough to trusted ui that it probably warrants its own section heading.

File: index.bs [hunk 9]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
 
 {{XRSession}}s MUST have their [=visibility state=] set to {{XRVisibilityState/"hidden"}} when the user is interacting with potentially sensitive UI from the user agent (such as entering a URL) in the trusted environment.
 
+<section class="unstable">
+Gaze Tracking {#gazetracking-security}
+-------------
+
+While the API does not yet expose eye tracking capabilities a lot can be inferred about where the user is looking by tracking the orientation of their head. This is especially true of XR devices that have limited input capabilities, such as Google Cardboard, which frequently require users to control a "gaze cursor" with their head orientation. This means that it may be possible for a malicious page to infer what a user is typing on a virtual keyboard or how they are interacting with a virtual UI based solely on monitoring their head movements. For example: if not prevented from doing so a page could estimate what URL a user is entering into the user agent's URL bar.
```
Comment (NellWaliczek) [inline index.bs]: Is this text still helpful?

>While the API does not yet expose eye tracking capabilities a lot can be inferred about where the user is looking by tracking the orientation of their head. This is especially true of XR devices that have limited input capabilities, such as Google Cardboard, which frequently require users to control a "gaze cursor" with their head orientation.

Comment (NellWaliczek) [inline index.bs]: This text can probably be repurposed as the intro paragraph if it is adjusted slightly.

>This means that it may be possible for a malicious page to infer what a user is typing on a virtual keyboard or how they are interacting with a virtual UI based solely on monitoring their head movements. For example: if not prevented from doing so a page could estimate what URL a user is entering into the user agent's URL bar.

File: index.bs [hunk 10]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+
+To prevent this risk the user agent MUST set the [=visibility state=] of all {{XRSession}}s to {{XRVisibilityState/"hidden"}} when the user is interacting with sensitive, trusted UI such as URL bars or system dialogs. Additionally, to prevent a malicious page from being able to monitor input on other pages the user agent MUST set the {{XRSession}}'s [=visibility state=] to {{XRVisibilityState/"hidden"}} if the [=currently focused area=] does belong to the document which created the {{XRSession}}.
```
Comment (NellWaliczek) [inline index.bs]: We should make it clear that this applies to both types of trusted UI

File: index.bs [hunk 11]
```diff
@@ -2214,25 +2214,47 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+
+Additionally, page content has the ability to make users uncomfortable in ways not related to performance. Badly applied tracking, strobing colors, and content intended to offend, frighten, or intimidate are examples of content which may cause the user to want to quickly exit the XR experience. Removing the XR device in these cases may not always be a fast or practical option. To accommodate this the user agent SHOULD provide users with an action, such as pressing a reserved hardware button or performing a gesture, that escapes out of WebXR content and displays the user agent's [=trusted UI=].
 
 {{XRSession}}s MUST have their [=visibility state=] set to {{XRVisibilityState/"hidden"}} when the user is interacting with potentially sensitive UI from the user agent (such as entering a URL) in the trusted environment.
 
+<section class="unstable">
+Gaze Tracking {#gazetracking-security}
```
Comment (NellWaliczek) [inline index.bs]: I'm not sure this actually needs to be called out as gaze tracking.  It seems like it can be merged into the previous section more deeply.

File: index.bs (hunk 3 above)
Comment (Manishearth) [reply index.bs]: Yes

**Comment by Manishearth** [comment]
Comments addressed.

File: index.bs [hunk 12]
```diff
@@ -2214,24 +2214,41 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+A <dfn>Trusted UI</dfn> is an interface that the user can trust comes from the user agent, which the user may interact with without interference from the page. The user agent MUST support showing a [=trusted UI=]. Some form of [=trusted UI=] MUST be used to show permissions prompts.
+
+
+Broadly speaking, there are two options for user agents who wish to support [=trusted UI=]. One option is the <dfn>trusted immersive UI</dfn>, which is a [=trusted UI=] which does not exit immersive mode. It is tricky to design a good [=trusted immersive UI=] since the page can effectively draw any pixels it wishes to. User agents are not required to support [=trusted immersive UI=], they may instead temporarily pause/exit immersive mode and show non-immersive [=trusted UI=] to the user.
```
Comment (NellWaliczek) [inline index.bs]: nit: it's not entirely clear why "drawing any pixels" is a problem. I'm not super thrilled with the exact phrasing of this suggestion, but maybe something kinda like...

```suggestion
Broadly speaking, there are two options for user agents who wish to support [=trusted UI=]. One option is <dfn>trusted immersive UI</dfn>, which is a [=trusted UI=] which does not exit immersive mode. Implementing [=trusted immersive UI=] can be challenging because `XRWebGLLayer` buffers fill the XR Device display and the User Agent does not typically "reserve" pixels for its own use. User agents are not required to support [=trusted immersive UI=], they may instead temporarily pause/exit immersive mode and show non-immersive [=trusted UI=] to the user.
```

File: index.bs [hunk 13]
```diff
@@ -2214,24 +2214,41 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+</div>
+
+In some cases it may be possible for a malicious page to infer what a user is typing on a virtual keyboard or how they are interacting with a virtual UI based solely on monitoring their head movements. This is especially true on devices that have limited input capabilities, such as Google Cardboard, which frequently require users to control a "gaze cursor" with their head orientation. For example: if not prevented from doing so a page could estimate what URL a user is entering into the user agent's URL bar by monitoring the user's interaction with the keyboard.
+
+To prevent this risk the user agent MUST set the [=visibility state=] of all {{XRSession}}s to {{XRVisibilityState/"hidden"}} when the user is interacting with sensitive, [=trusted UI=] ([=trusted immersive ui|immersive=] or non-immersive) such as URL bars or system dialogs. Additionally, to prevent a malicious page from being able to monitor input on other pages the user agent MUST set the {{XRSession}}'s [=visibility state=] to {{XRVisibilityState/"hidden"}} if the [=currently focused area=] does belong to the document which created the {{XRSession}}.
```
Comment (NellWaliczek) [inline index.bs]: Nit: we can probably strike the word "sensitive"

File: index.bs [hunk 14]
```diff
@@ -2214,24 +2214,41 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+Note: Examples of [=trusted UI=] include:
+  - The default 2D mode browser shown when not in immersive mode
+  - A prompt shown within immersive mode which can only be interacted with via a reserved hardware button to prevent spoofing
+  - Pausing the immersive session and showing some form of native system environment in which a prompt can be shown
 
-When navigating between pages in XR the user agent should display trusted UI elements informing the user of the security information of the site they are navigating to which is normally presented by the 2D UI, such as the URL and encryption status.
+</div>
+
+In some cases it may be possible for a malicious page to infer what a user is typing on a virtual keyboard or how they are interacting with a virtual UI based solely on monitoring their head movements. This is especially true on devices that have limited input capabilities, such as Google Cardboard, which frequently require users to control a "gaze cursor" with their head orientation. For example: if not prevented from doing so a page could estimate what URL a user is entering into the user agent's URL bar by monitoring the user's interaction with the keyboard.
```
Comment (NellWaliczek) [inline index.bs]: I think this whole paragraph can probably be replaced with one sentence at the beginning of the next paragraph that says "Input snooping is bad, yo" because it isn't just gaze-based input that can be sniffed.  tracked pointers and screen are just as susceptible.

File: index.bs [hunk 15]
```diff
@@ -2214,24 +2214,41 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
 
 Trusted Environment {#trustedenvironment-security}
 -------------------
 
-If the virtual environment does not consistently track the user's head motion with low latency and at a high frame rate the user may become disoriented or physically ill. Since it is impossible to force pages to produce consistently performant and correct content the user agent MUST provide a tracked, trusted environment and an [=XR Compositor=] which runs asynchronously from page content. The compositor is responsible for compositing the trusted and untrusted content. If content is not performant, does not submit frames, or terminates unexpectedly the user agent should be able to continue presenting a responsive, trusted UI.
+A <dfn>Trusted UI</dfn> is an interface that the user can trust comes from the user agent, which the user may interact with without interference from the page. The user agent MUST support showing a [=trusted UI=]. Some form of [=trusted UI=] MUST be used to show permissions prompts.
```
[the remark this answers is not in the document]
Comment (NellWaliczek) [reply index.bs]: On second thought, maybe remove the part about permissions prompts and put it down below when we talk about the reasons trusted UI should be displayed.

File: index.bs [hunk 16]
```diff
@@ -2214,24 +2214,41 @@ Note: Is is suggested that poses reported relative to a {{XRReferenceSpaceType/"
...
+
+Additionally, page content has the ability to make users uncomfortable in ways not related to performance. Badly applied tracking, strobing colors, and content intended to offend, frighten, or intimidate are examples of content which may cause the user to want to quickly exit the XR experience. Removing the XR device in these cases may not always be a fast or practical option. To accommodate this the user agent SHOULD provide users with an action, such as pressing a reserved hardware button or performing a gesture, that escapes out of WebXR content and displays the user agent's [=trusted UI=].
```
Comment (NellWaliczek) [inline index.bs]: should SHOULD be MUST?

Comment (Manishearth) [reply index.bs]: It was SHOULD in the old text, but I made it a MUST.

**Comment by Manishearth** [comment]
Addressed.

**Comment by Manishearth** [comment]
Thanks everyone!
